db/snapshot/backup_task.cc

```diff
@@ -103,51 +108,139 @@ future<> backup_task_impl::upload_component(sstring name) {
     }
 }
 
+future<> backup_task_impl::list_snapshot_dir() {
+    auto snapshot_dir_lister = directory_lister(_snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
+
+    try {
+        snap_log.debug("backup_task: listing {}", _snapshot_dir.native());
+        while (auto component_ent = co_await snapshot_dir_lister.get()) {
+            const auto& name = component_ent->name;
+            auto file_path = _snapshot_dir / name;
+            auto st = co_await file_stat(file_path.native());
+            try {
+                auto desc = sstables::parse_path(file_path, "", "");
+                _sstable_comps[desc.generation].emplace_back(name, st.size);
+                ++_num_sstable_comps;
+
+                if (desc.component == sstables::component_type::Data) {
+                    // If the sstable is already unlinked after the snapshot was taken
+                    // track its generation in the unlinked_sstables list
+                    // so it can be prioritized for backup
+                    if (st.number_of_links == 1) {
+                        snap_log.trace("do_backup: sstable with gen={} is already unlinked", desc.generation);
+                        _unlinked_sstables.push_back(desc.generation);
+                    }
+                }
+            } catch (const sstables::malformed_sstable_exception&) {
+                _non_sstable_files.emplace_back(name, st.size);
+            }
+        }
+        snap_log.debug("backup_task: found {} SSTables consisting of {} component files, and {} non-sstable files",
+            _sstable_comps.size(), _num_sstable_comps, _non_sstable_files.size());
+    } catch (...) {
+        _ex = std::current_exception();
+        snap_log.error("backup_task: listing {} failed: {}", _snapshot_dir.native(), _ex);
+    }
+
+    co_await snapshot_dir_lister.close();
+    if (_ex) {
+        co_await coroutine::return_exception_ptr(std::move(_ex));
+    }
+}
+
+future<> backup_task_impl::backup_file(const comp_desc& desc) {
+    _as.check();
+
+    auto gh = _uploads.hold();
+    uint64_t required = std::min(desc.size, max_concurrency);
+    auto units = co_await get_units(_concurrency_sem, required, _as);
+
+    // okay to drop future since uploads is always closed before exiting the function
+    std::ignore = upload_component(desc.name).handle_exception([this] (std::exception_ptr e) {
```

Comment:
do we have a doc with the layout format in s3 for backup? I wonder if we're not leveraging the concurrency provided by using different prefixes.

Reply:
For keyspace storage options that format is documented (maybe poorly) in docs/dev/object_storage.md
For backup, the layout is here: https://manager.docs.scylladb.com/stable/backup/specification.html